dex/meter/meterer.go

```diff
@@ -0,0 +1,72 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package meter
+
+import (
+	"sync"
+	"time"
+)
+
+// SignalMeterer meters a simple signal to prevent tight grouping. The user
+// specifies a minimum delay (minDelay). If a signal arrives within minDelay of
+// the previous signal, that signal will be scheduled for later to respect the
+// configured delay. If multiple signals arrive within minDelay, they will be
+// grouped into a single delayed signal. Error signals are never delayed.
+type SignalMeterer struct {
+	minDelay time.Duration
+	recv     func(error)
+
+	mtx       sync.Mutex
+	last      time.Time
+	scheduled *time.Timer
+}
+
+// NewSignalMeterer creates a new SignalMeterer.
+func NewSignalMeterer(minDelay time.Duration, recv func(error)) *SignalMeterer {
+	return &SignalMeterer{
+		minDelay: minDelay,
+		recv:     recv,
+	}
+}
+
+// Signal sends a signal, with a delay if appropriate.
+func (s *SignalMeterer) Signal(err error) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	// Errors are passed immediately.
+	if err != nil {
+		go s.recv(err)
```

Comment:
I think you might as well replace `recv  func(error)` with `recv  chan error`, and doing `s.recv <- err` instead of `go s.recv(err)`.

The goroutine allocations and runtime overhead seem avoidable, and it feels tailored for a channel anyway.  I get the sense you're using goroutines as a way to use the runtime for effectively unlimited buffering, which could backfire in addition to being a bit inefficient.

Plus in `Swapper`, I think a channel is more natural since you've got a select loop going already.  illustration: https://github.com/decred/dcrdex/commit/124ec646226926a7b3a70ee290751ba90fb5fc8c

And as usual lately I'm looking at patterns that eliminate locks and fields with event loops and closure, so I'm toying with this alternate mechanism: https://github.com/decred/dcrdex/commit/817b80063fe33f491999155ae21c27ab148279c9
Just toying around, don't feel like you need to take either suggestion.

Reply:
https://github.com/decred/dcrdex/commit/817b80063fe33f491999155ae21c27ab148279c9 is beautiful 😍.